Design.md

```diff
@@ -136,46 +121,90 @@ This API will update the existing storage tenant with the provided `cd_tenant_id
+* `assumeRoleCache` cache will be a **[key, value]** pair of **[Role_Arn, Temporary_Credentials]** respectively.
+    * Each `assumeRoleCache` entry will have a 50-minute TTL (its session token is valid only for 60 minutes).
+    * When accessing credentials on the cache, refresh the credentials as well as `needsRefresh` (using AWS SDK `refresh()`), as they are subject to expiration.
+    * Cache design can be found [here](#Cache-Design).
+    
 ### Create User
 This API creates a user on Vault.
 1. `create-user` api will be called using assumed role credentials.
-    * `cd_user_id` is stored as username.
-1. `attach-policy` for the new user will be called using the assumed role credentials with full S3 and IAM access permissions.
+    * The tenant user's `cdUserId` is stored as `username` in the corresponding Vault user.
+    * The tenant user's `username`, `role` enum value, `emailAddress` and `cdTenantID` are stored in the Vault user path as `/<tenantUsername>/<roleEnumValue>/<email>/<cdTenantID>/`.
+1. `get-policy` api will be called with policy name `userPolicy@[account-id]` using assumed role credentials
+1. If the `get-policy` API does not return any policy, using the assumed role credentials create an IAM managed policy with policy name `userPolicy@[account-id]` with full S3 and KMS access.
+1. `attach-policy` for the new user will be called using the assumed role credentials with the policy arn as `arn:aws:iam::[account-id]:policy/userPolicy@[account-id]`.
 1. `create-s3credentials` api will be called using assumed role credentials
 
 ### Create User Example Activity Diagram
 ![Create User](Design_Files/OSE-OSIS-Create-User.png)
 
-### Query Users
-This API will query users on Vault using a `filter` parameter.
-1. `list-users` api will be called using assumed role credentials.
-1. Filter the result users using the `filter` parameter in the request
-
 ### List Users
 This API will list users on Vault.
+1. The `list-users` API can be called with following parameters:
+    * `offset`: The start index of tenants to return (optional)
+    * `limit`: The maximum number of tenants to return (optional)
 1. `list-users` api will be called using assumed role credentials.
+    * If `offset` is present, return the `list-users` result by providing the `offset` value as the `marker` parameter.
+
+### Query Users
+This API will query users on Vault using a `filter` parameter with the filters of `cd_tenant_id` and `display_name`.
+1. Extract the `cd_tenant_id` filter from the `filter` parameter to call `list-accounts` using vaultclient to retrieve the `accountID`. 
+1. Extract the `displayname` value from the `filter` parameter.
+1. `list-users` api will be called using assumed role credentials with the `path-prefix` as `/<display_name>/`.
```

Comment:
Again, it would be nice to know who is calling the`list-users` API, in which case, "Vault? F00? will call the `list-users` API using ..."

```suggestion
1. The `list-users` API will be called using assumed role credentials with the `path-prefix` as `/<display_name>/`.
```